File: blueluak-jse/src/test/java/net/blueva/luak/LoadOrderTest.java
/******************************************************************************
 *  ____  _            _                _  __
 * | __ )| |_   _  ___| |   _   _  __ _| |/ /
 * |  _ \| | | | |/ _ \ |  | | | |/ _` | ' /
 * | |_) | | |_| |  __/ |__| |_| | (_| | . \
 * |____/|_|\__,_|\___|_____\__,_|\__,_|_|\_\
 *
 *  BlueLuaK
 *  https://github.com/BluevaDevelopment/BlueLuaK
 *
 *  Based on LuaJ (https://luaj.org)
 *  Original work Copyright (c) 2009 Luaj.org
 *  Modifications Copyright (c) 2026 Blueva Development
 *
 *  SPDX-License-Identifier: MIT
 ******************************************************************************/
package net.blueva.luak;

import java.io.InputStream;
import java.io.Reader;

import junit.framework.TestCase;

import net.blueva.luak.lib.jse.JsePlatform;
import net.blueva.luak.server.Launcher;
import net.blueva.luak.server.LuajClassLoader;

// Tests using class loading orders that have caused problems for some use cases.
public class LoadOrderTest extends TestCase {

	public void testLoadGlobalsFirst() {
		Globals g = JsePlatform.standardGlobals();
		assertNotNull(g);
	}

	public void testLoadStringFirst() {
		LuaString BAR = LuaString.valueOf("bar");
		assertNotNull(BAR);
	}

	public static class TestLauncherLoadStringFirst implements Launcher {
		// Static initializer that causes LuaString->LuaValue->LuaString
		private static final LuaString FOO = LuaString.valueOf("foo");

		public Object[] launch(String script, Object[] arg) {
			return new Object[] { FOO };
		}

		public Object[] launch(InputStream script, Object[] arg) {
			return null;
		}

		public Object[] launch(Reader script, Object[] arg) {
			return null;
		}
	}

	public void testClassLoadsStringFirst() throws Exception {
		Launcher launcher = LuajClassLoader
				.NewLauncher(TestLauncherLoadStringFirst.class);
		Object[] results = launcher.launch("foo", null);
		assertNotNull(results);
	}

}

File: luak-jvm/src/test/kotlin/org/basaltmc/luak/LoadOrderTest.kt
/******************************************************************************
 *  _                _
 * | |   _   _  __ _| | __
 * | |  | | | |/ _` | |/ /
 * | |__| |_| | (_| |   <
 * |_____\__,_|\__,_|_|\_\
 *
 *  Luak
 *  https://github.com/BasaltProject/Luak
 *
 *  Based on LuaJ (https://luaj.org)
 *  Original work Copyright (c) 2009 Luaj.org
 *  Modifications Copyright (c) 2026 Basalt Project
 *
 *  SPDX-License-Identifier: MIT
 ******************************************************************************/
package org.basaltmc.luak

import junit.framework.TestCase
import org.basaltmc.luak.lib.jvm.JvmPlatform.standardGlobals
import org.basaltmc.luak.server.Launcher
import org.basaltmc.luak.server.LuaClassLoader
import java.io.InputStream
import java.io.Reader

// Tests using class loading orders that have caused problems for some use cases.
class LoadOrderTest : TestCase() {
    fun testLoadGlobalsFirst() {
        val g = standardGlobals()
        assertNotNull(g)
    }

    fun testLoadStringFirst() {
        val BAR = LuaString.valueOf("bar")
        assertNotNull(BAR)
    }

    class TestLauncherLoadStringFirst : Launcher {
        override fun launch(script: String?, arg: Array<Any?>?): Array<Any?>? {
            return arrayOf<Any?>(FOO)
        }

        override fun launch(script: InputStream?, arg: Array<Any?>?): Array<Any?>? {
            return null
        }

        override fun launch(script: Reader?, arg: Array<Any?>?): Array<Any?>? {
            return null
        }

        companion object {
            // Static initializer that causes LuaString->LuaValue->LuaString
            private val FOO = LuaString.valueOf("foo")
        }
    }

    @Throws(Exception::class)
    fun testClassLoadsStringFirst() {
        val launcher = LuaClassLoader
            .NewLauncher(TestLauncherLoadStringFirst::class.java)
        val results = launcher.launch("foo", null)
        assertNotNull(results)
    }
}
